TropicalGeometry: groebner_fan.jl, new implementation for Groebner fans

YueRen opened the pull request:
Main new command:
- groebner_fan

Secondary new commands:
- maximal_groebner_cone
- homogeneity_space
- homogeneity_vector

@martapanizzut @ederc Would you be so kind and review this pull request please? I'm happy to give you a quick overview of the code over Zoom if you want.

YueRen committed:
TropicalGeometry: groebner_fan.jl, new implementation for Groebner fans

TropicalGeometry: groebner_fan.jl, typo fixes suggested by ederc

Co-authored-by: ederc <ederc@mathematik.uni-kl.de>

TropicalGeometry: groebner_fan.jl, consistency and wording fixes by fingolfin

Co-authored-by: Max Horn <max@quendi.de>

TropicalGeometry: added test for groebner_fan

remaining functions in groebner_fan.jl are well covered by
documentation examples

TropicalGeometry: maximal_groebner_cone, sanitised input type

new input type is return type of `groebner_basis`

TropicalGeometry: moved ordering(::Oscar.IdealGens) to mpoly.jl

TropicalGeometry: groebner_fan.jl, added empty lines between functions

Co-authored-by: Max Horn <max@quendi.de>

TropicalGeometry: groebner_fan.jl, vertical spacing

TropicalGeometry: Update src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl

Co-authored-by: Max Horn <max@quendi.de>

TropicalGeometry: Update src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl

Co-authored-by: Max Horn <max@quendi.de>

TropicalGeometry: groebner_fan.jl, fixed double definition

TropicalGeometry: groebner_fan.jl, fixed doctests

TropicalGeometry: groebner_fan.jl, replaced PolyhedralFan

YueRen commented:
Hm, any introductory tips for what I must do to make codecov bot happy?

benlorenz commented:
The coverage went down because the doctests failed and most of the failures I found are because the code uses `contains(Cone, Vector)` which was deprecated recently:
```
ERROR: `contains(C::Cone, v::AbstractVector)` is deprecated, use `Base.in(v, C)` instead.
│    Stacktrace:
│     [1] depwarn(msg::String, funcsym::Symbol; force::Bool)
│       @ Base ./deprecated.jl:82
│     [2] depwarn(msg::String, funcsym::Symbol)
│       @ Base ./deprecated.jl:80
│     [3] contains(C::Cone{QQFieldElem}, v::Vector{ZZRingElem})
│       @ Oscar ./deprecated.jl:71
│     [4] homogeneity_vector(G::Vector{QQMPolyRingElem})
│       @ Oscar ~/work/Oscar.jl/Oscar.jl/src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:123
│     [5] top-level scope
│       @ none:1
```
Just change these to `v in C`.

YueRen commented:
Ah, got it.  I fixed it and changed the test whether `homogeneityVector` is inside the positive orthant from  `homogeneityVector>zeros(fmq,length(homogeneityVector))` to `isnothing(findfirst(vi -> (vi<=0), homogeneityVector))` which should be minimally faster. I also rebased it to the latest master just to make sure that all my examples still work.

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+@doc raw"""
+    homogeneity_vector(G::Vector{<:MPolyElem})
+
+If it exists, a positive `Vector{fmpz}` under which every of `G` is weighted homogeneous.  Otherwise, `nothing`.
```
```suggestion
If it exists, a positive `Vector{fmpz}` under which every element of `G` is weighted homogeneous.  Otherwise, `nothing`.
```

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+If it exists, a positive `Vector{fmpz}` under which every of `G` is weighted homogeneous.  Otherwise, `nothing`.
+
+# Note
+Suppose `G` is a reduced Groebner basis of an ideal `I` with respect to a global ordering. If a `Vector{fmpz}` is returned, then `I` will be weighted homogeneous with respect to it.  If `nothing` is returned, then `I` is not weighted homogeneous with respect to any positive weight vector.
```
```suggestion
Suppose `G` is the reduced Groebner basis of an ideal `I` with respect to a global ordering. If a `Vector{fmpz}` is returned, then `I` is weighted homogeneous with respect to it.  If `nothing` is returned, then `I` is not weighted homogeneous with respect to any positive weight vector.
```

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+    maximal_groebner_cone(G::Vector{<:MPolyElem}, ord::MonomialOrdering, homogeneityWeight::Union{Nothing,Vector{fmpz}}=nothing)
+
+# Input
+- `G`: A reduced Groebner basis of an ideal `I` with respect to `ord`,
```
```suggestion
- `G`: The reduced Groebner basis of an ideal `I` with respect to `ord`,
```

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+# Input
+- `G`: A reduced Groebner basis of an ideal `I` with respect to `ord`,
+- `ord`: A global monomial ordering,
+- `homogeneityWeight`: Either `Nothing` or a positive weight vector under which `I` is weighted homogeneous.
```
```suggestion
- `homogeneityWeight`: Either `nothing` or a positive weight vector under which `I` is weighted homogeneous.
```

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+@doc raw"""
+    homogeneity_vector(G::Vector{<:MPolyElem})
+
+If it exists, a positive `Vector{fmpz}` under which every of `G` is weighted homogeneous.  Otherwise, `nothing`.
```
Wouldn't it make more sense to return a zero vector instead of `nothing` if no such homgeneity vector exists?

YueRen replied:
That is very true.  I guess I just wanted to play around with type dispatch and use it to distinguish between the (quasi-)homogeneous and the non-homogeneous case.

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+    # calling with a positive homogeneityWeight
+    # means that C does not be restricted to the positive orthant
+    return maximal_groebner_cone_extended(G,ord)
+end
```
Here I am again have to ask: Instead of `nothing` wouldn't it be easier to have an all zeroes vector?

fingolfin replied:
In which sense would it be easier?

(Don't get me wrong, I am not opposed, I'd just like to understand the argument).

I see pros and cons to both. But at least `nothing` makes it less likely that one accidentally uses a zero vector as if it was a weight vector?

The downside of `nothing` I can think of is that one has to be careful to preserve type stability. But Julia does have provisions for that: if one branches based on the return type then Julia is able to deduce that correctly. E.g. consider this example:
```
julia> f(x) = x > 0 ? x : nothing
f (generic function with 1 method)

julia> function g(x)
         y = f(x)
         y === nothing && return 0
         return y + 1
       end
g (generic function with 1 method)

julia> @code_warntype g(7)
MethodInstance for g(::Int64)
  from g(x) in Main at REPL[13]:1
Arguments
  #self#::Core.Const(g)
  x::Int64
Locals
  y::Union{Nothing, Int64}
Body::Int64
1 ─      (y = Main.f(x))
│   %2 = (y === Main.nothing)::Bool
└──      goto #3 if not %2
2 ─      return 0
3 ─ %5 = (y::Int64 + 1)::Int64
└──      return %5
```
Here initially the type for `y` is inferred as `Union{Nothing, Int64}`. But after the `y === nothing && return 0` Julia is clever enough to deduce that now `y` really is an `Int`. This also works when using `isnothing(y)` or `y isa Nothing`.

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+
+
+
+# Returns an interior points on each facet.
```
```suggestion
# Returns an interior point on each facet, respectively.
```

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+# Requires a monomial ordering that is compatible with respect to u, i.e., the leading monomial is of highest weighted degree.
+function initial(g::MPolyElem, ordering::MonomialOrdering, u::Vector{fmpz})
+
+    lt,tail = Iterators.peel(terms(g,ordering=ordering));
```
There is also `leading_term()` and `tail()` available in Oscar.

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+# Assumes:
+# - Hnew is a Groebner basis of the initial ideal w.r.t. ordH
+# - G is a Groebner basis of the original ideal w.r.t. ordG
+# - ordH and ordG are adjacent orderings
```
Is the adjacency of `ordG` and `ordH` tested in `groebner_lift`?

YueRen replied:
No, it is not.  I left it out because it requires some non-trivial computations and it is not necessary.

To test the adjacency under the assumptions that `Hnew` and `G` are Groebner bases w.r.t. `ordH` and `ordG`, respectively, one would need to verify
1. that there is a weight vector `w` with `initial_w(G)=Hnew` (as sets of polynomials),
2. `lead(G,ordG)` and `lead(G,ordH)` are not the same (as sets of polynomials).

Point 1 implies that `ordG` and `ordH` are either adjacent or the same, Point 2 implies that they are not the same. (Note that "adjacent" here means that the Groebner cones share a common face containing `w`, but that face need not be one-codimensional)

ederc commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+# - u is an interior facet point of the maximal Groebner cone around ordG
+# - v is an outer normal vector of the facet containing u
+# Returns:
+# - a reduced Groebner basis with respect to the adjacent ordering
```
```suggestion
# - the reduced Groebner basis with respect to the adjacent ordering
```

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+@doc raw"""
+    homogeneity_space(G::Vector{<:MPolyElem})
+
+A `Cone` that is the space of all weight vectors under which all elements of `G` are weighted homogeneous.  The cone will be a linear subspace without any rays.
```
```suggestion
Return a `Cone` that is the space of all weight vectors under which all elements of `G` are weighted homogeneous.  The cone will be a linear subspace without any rays.
```

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+A `Cone` that is the space of all weight vectors under which all elements of `G` are weighted homogeneous.  The cone will be a linear subspace without any rays.
+
+# Note
+If `G` is a reduced Groebner basis of an ideal `I` with respect a global ordering, then the returned cone is the space of all weight vectors with respect to which `I` is weighted homogeneous.
```
Please avoid lines which are too long (say < 100 chars, though as a rule of thumb I use between 78 and 92 lines). This helps with reviewing changes to these lines in GitHub and via `git` diff. It's just reality that if later we need to fix a typo in a line with for example 300 characters, it is a hassle to review this. 

Separately: maybe use an [admonition](https://docs.julialang.org/en/v1/stdlib/Markdown/#Admonitions) for notes.

```suggestion
!!! note
    If `G` is a reduced Groebner basis of an ideal `I` with respect a global
    ordering, then the returned cone is the space of all weight vectors with
    respect to which `I` is weighted homogeneous.
```

YueRen replied:
Quick question: will the line breaks that I enter here manually cause line breaks in the documentation?

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+@doc raw"""
+    homogeneity_vector(G::Vector{<:MPolyElem})
+
+If it exists, a positive `Vector{fmpz}` under which every element of `G` is weighted homogeneous.  Otherwise, `nothing`.
```
```suggestion
Return a positive `Vector{fmpz}` under which every element of `G` is weighted
homogeneous. If no such vector exists, return `nothing`.
```

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+ x3 + 1
+
+julia> homogeneity_vector(G3)
+
```
Should there be a return value here?

YueRen replied:
No, the return is `nothing`, but I can see how that can be confusing, I will add a `== nothing` (which will then be `true`) to make it clearer.

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+- `verbose_level`: If positive, print out how many cones have been computed during the traversal
+
+# Output
+A `PolyhedralFan` that is the Groebner fan of `I`.
```
This Input/Output format is discouraged by Julia, and I agree with that.  It should ideally only be used if it just can't be reasonably avoided. But here I think it can easily be avoided? E.g. like this:
```suggestion
Return a `PolyhedralFan` representing the Groebner fan of `I`, where is a multivariate polynomial ideal over a field.

If `verbose_level` is positive, also print how many cones have been computed during the traversal.
```
The description of the `return_*` kwargs seems to be redundant with what follows?

BTW also `verbose_level` should perhaps be removed and instead replaced with use of `@vprint` and friends -- but that could be done in a future PR, of course

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,544 @@
+
+If `I` is not weighted homogeneous with respect to a positive weight vector, the Groebner fan will be restricted to the positive orthant.  Otherwise, the Groebner fan will span the entire space and have a non-trivial lineality space.
+
+If `return_groebner_bases==true`, returns a dictionary whose keys are interior points of the maximal cones and whose values are the Groebner bases for those cones.  Their union will be a universal Groebner basis.
```
My understanding is that the dictionary is returned *in addition* to the fan, is that right? I'd suggest to make this clear, one way or the other. E.g.:
```suggestion
If `return_groebner_bases==true`, also return a dictionary whose keys are interior points of the maximal cones and whose values are the Groebner bases for those cones.  Their union will be a universal Groebner basis.
```

fingolfin reviewed (commented):
I can't comment on the math in here, but I have some nitpicky stylistic comments. They are not meant to annoy you, @YueRen (I apologies if they do), but rather aim at improving consistency on the long term -- in that spirit I'd rather "complain" early to give you chance to react and adapt. But of course I understand this code is under development, so that perhaps you meant to change some of it anyway. Also it's also OK to do some or all of this later.

Finally, don't hesitate to question any of my suggestions :-)

thofma commented:
The interface is for `(::Vector{<: MPolyElem}, ::MonomialOrdering)`, but the docstrings speak of Gröbner bases. But in Oscar land these are not considered to be Gröbner bases. Is there a reason why the input is not the type of a Gröbner basis?

thofma commented:
I would suggest we add some tests? This would also make codecov happy.

YueRen commented:
@thofma 

> The interface is for (::Vector{<: MPolyElem}, ::MonomialOrdering), but the docstrings speak of Gröbner bases. But in Oscar land these are not considered to be Gröbner bases. Is there a reason why the input is not the type of a Gröbner basis?

I've changed the "official" input type of `Oscar.IdealGens` now, which is what `groebner_basis` returns (though internally I'm still using the old version).

> I would suggest we add some tests? This would also make codecov happy.

I added two more examples for groebner_fan as tests (which take 4 and 6 seconds, respectively), the remaining functions are already well-covered by their examples in the documentation.



The pull-request is almost good to go. Two things are missing:
1. I needed and hence added the function `ordering(G::Oscar.IdealGens) = G.ord`.  But `groebner_basis.jl` is definitely not the right place for it.  Where should I put it?
2. Squashing all commits before the merge.

thofma commented:
Thanks. I think this needs a rebase before CI runs again.

YueRen commented:
It's rebased onto the current master and I've squashed all commits.  As soon as all tests are passed, it's ready to be merged.

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,541 @@
+    homogeneityWeight = homogeneity_vector(G)
+    return maximal_groebner_cone(G,ord,homogeneityWeight)
+end
+function maximal_groebner_cone(G::Oscar.IdealGens{<:MPolyElem}; homogeneityWeight::Vector{fmpz}=fmpz[])
```
This is a minor nitpick, and not a merge blocker, but: please put an empty line between the end of one function and the start of the next
```suggestion

function maximal_groebner_cone(G::Oscar.IdealGens{<:MPolyElem}; homogeneityWeight::Vector{fmpz}=fmpz[])
```

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,541 @@
+        G[i] = reduce(G[i],G[1:i-1],ordering=ord)
+    end
+    return G
+
```
Another minor nitpick, not a blocker: it seems unusual to leave this empty line between `return` and `end`

YueRen replied:
Fixed in the latest commit.

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,541 @@
+                                         interior_facet_point::Vector{fmpz},
+                                         outer_normal_vector::Vector{fmpz})
+
+    return weight_ordering(Int.(homogeneityVector),
```
Wait, so does that mean all those `Vector{fmpz}` can be safely converted to `Vector{Int}`? So why don't we store them as `Vector{Int}` in the first place?

YueRen replied:
A lot of objects in the tropical code lives in the intersection between polymake and singular.  The former uses `fmpz` and the latter uses `int`.   Since changing polymake to `int` makes less sense to me than teaching Singular to use `fmpz`, I decided to keep everything as `fmpz` and convert to `int` only when necessary.  It's easier to remember than deciding on a case-by-case basis when to use `fmpz` and when to use `int`.

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,541 @@
+    end
+
+    # otherwise return what was required
+    output = Any[]
```
Use of `Any` here is redundant.
```suggestion
    output = []
```

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,541 @@
+    ###
+    while !isempty(workingList)
+        if verbose_level>0
+            println("#workingList / #finishedList: ",length(workingList)," / ",length(finishedList))
```
You may want to look into using `@vprint` here instead (in a future revision of this code, not in this PR)

fingolfin commented on src/AlgebraicGeometry/TropicalGeometry/groebner_fan.jl:
```diff
@@ -0,0 +1,541 @@
+    #   Preference is given to the ones vector for the sake of optimisation
+    ###
+    G = groebner_basis(I,complete_reduction=true)
+    ord = G.ord
```
Presumably here you want to use your new `ordering` method?
```suggestion
    ord = ordering(G)
```

YueRen commented:
Okay, the pull request is ready to be merged:
- I commit the changes suggested by Max on Github
- I changed the empty new lines to be more consistent with the rest of Oscar
- I rebased to the latest version of the master branch
- I squashed all commits into one

I will be merging it oncethe tests complete.

thofma commented:
I restarted so that the coverage (should) work.

YueRen commented:
Tests are passing now, if somebody could merge the pull request that would be much appreciated!

YueRen commented:
Tests are passing now, if somebody could merge the pull request that would be much appreciated!
